game.js

```diff
@@ -63,6 +74,93 @@ document.querySelector('.restart').addEventListener('click', () => {
         cell.classList.remove('disabled', 'x', 'o')
     })
 
+    game.xTurn = true
+    game.xState = []
+    game.oState = []
+})
+
+document.querySelector('.return-one-step').addEventListener('click', () => {  
+    const xLength = game.xState.length;
+    const oLength = game.oState.length;
+    const lastXCell = document.querySelector(`[data-value = '${game.xState[xLength -1]}']`)
+    const lastOCell = document.querySelector(`[data-value = '${game.oState[oLength -1]}']`)
+ if (xLength > 0){
+    if (game.xTurn) { 
+     lastOCell.classList.remove('o', 'disabled');
+     game.oState.pop();
+    game.xTurn = false;
+ }
+ else{
+     lastXCell.classList.remove('x', 'disabled');
+     game.xState.pop();
+     game.xTurn = true;
+ }}
+ })
+
+ document.querySelector('.show-record').addEventListener('click', () => {
+ const scoresArr = [];
+
+ for (const score in game.winnings) {
+  scoresArr.push([score, game.winnings[score]]);
+ }
+ if (scoresArr.length > 0){
+scoresArr.sort((a , b) =>  a[1] - b[1])
+alert(`The fastest game was game ${scoresArr[0][0]} with ${scoresArr[0][1]} moves`);
+}
+else {
+    alert("No winnings yet")
+}
+})
+
+document.querySelector('.save-game').addEventListener('click', function(){
+   if(game.xState.length){
+        game.savedGame.states.push(game.xState, game.oState);
+        game.savedGame.xTurn = game.xTurn;
+       this.disabled = true;
+    }
+    else{
+        alert("The board still empty")
+    }
+})
+
+document.querySelector('.load-game').addEventListener('click', () =>{
+if(game.savedGame.states.length > 0){
+        game.xState = game.savedGame.states[0];
+        game.oState = game.savedGame.states[1];
+
+ document.querySelectorAll('.grid-cell').forEach(function(cell){
+
+return game.xState.map((state) =>{
```

Comment:
- why use return ? return to who?
- why using map() did you need to create new array from the grid-cell? in case you just need to iterate over cell use forEach or for...of

Reply:
- I thought the return here is to the foreach function.
- I chose to use map() because I wanted to avoid nested loops.

Reply:
WDYM nested loops? the only different between map and forEach is that map return a new array (that you want to story it in variable and use it later and forEach is to make something forEach element in the array)

Reply:
OK, I need to clarify this. We talked about that before, when I used nested 'for' loops - the time complexity is n2.
So what I had in mind was that if I'll put a forEach inside forEach It'll manufacture the same result.
As I understand from your comment the nested forEach don't have the same problem.

Reply:
it is not related to time complexity (which is probably n^2 anyway , it doesn't meter if you have map loop inside forEach loop or nested forEach loops). it's related to the fact that map is not the correct function here ([see more](https://www.freecodecamp.org/news/4-main-differences-between-foreach-and-map/)).
about the time-complexity it's not big deal in this case because you have known n values (and it's not that big)

Reply:
I understand. After refactoring the code I went with forEach and includes() which also shortened the code into one forEach that contains x/oStates.includes